src/core/tests/focusManager.test.tsx

```diff
@@ -0,0 +1,106 @@
+import { sleep } from '../utils'
+import { focusManager } from '../focusManager'
+
+describe('focusManager', () => {
+  afterEach(() => {
+    // Reset removeEventListener private property to avoid side effects between tests
+    focusManager['removeEventListener'] = undefined
+  })
+
+  it('should call previous remove handler when replacing an event listener', () => {
+    const remove1Spy = jest.fn()
+    const remove2Spy = jest.fn()
+
+    focusManager.setEventListener(() => remove1Spy)
+    focusManager.setEventListener(() => remove2Spy)
+
+    expect(remove1Spy).toHaveBeenCalledTimes(1)
+    expect(remove2Spy).not.toHaveBeenCalled()
+  })
+
+  it('should use focused boolean arg', async () => {
+    let count = 0
+
+    const setup = (setFocused: (focused?: boolean) => void) => {
+      setTimeout(() => {
+        count++
+        setFocused(true)
+      }, 20)
+      return () => void 0
+    }
+
+    focusManager.setEventListener(setup)
+
+    await sleep(30)
+    expect(count).toEqual(1)
+    expect(focusManager.isFocused()).toBeTruthy()
+  })
+
+  it('should not notify listeners on focus if already focused', async () => {
+    const subscriptionSpy = jest.fn()
+    const unsubscribe = focusManager.subscribe(subscriptionSpy)
+
+    focusManager.setFocused(true)
+    expect(subscriptionSpy).toHaveBeenCalledTimes(1)
+    subscriptionSpy.mockReset()
+
+    focusManager.setFocused(false)
+    expect(subscriptionSpy).toHaveBeenCalledTimes(0)
+
+    unsubscribe()
+  })
+
+  it('should return true for isFocused if document is undefined', async () => {
+    const { document } = globalThis
+
+    // @ts-ignore
```

Comment:
let's please use `ts-expect-error` instead of `ts-ignore` wherever possible: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-9.html#-ts-expect-error-comments

Reply:
Done when possible in https://github.com/tannerlinsley/react-query/pull/2876/commits/ca22c597d27e0e4138f480f642ecbea374368b7c
Still have to use ` @ts-ignore ` to avoid `Unused '@ts-expect-error' directive.` in some cases.